### PR TITLE
APIS-1517 - Add Priv or ROPC app in gatekeeper

### PR DESCRIPTION
**Pull request by rose-garratt**

### CODE REVIEW

File: test/unit/services/ApplicationServiceSpec.scala [hunk 1]
```diff
@@ -22,12 +22,12 @@ import connectors.{ApiScopeConnector, ApplicationConnector}
...
 import model._
 import org.joda.time.DateTime
-import org.mockito.Mockito.{never, verify}
+import org.mockito.Mockito.{never, times, verify}
 import org.mockito.ArgumentCaptor
 import org.mockito.BDDMockito._
 import org.mockito.Matchers.{eq => mEq, _}
 import org.scalatest.mockito.MockitoSugar
-import services.ApplicationService
+import services.{ApplicationService}
```
Comment (peteslater-ee) [inline test/unit/services/ApplicationServiceSpec.scala]: Optimise imports.

File: app/controllers/ApplicationController.scala [hunk 2]
```diff
@@ -387,4 +387,58 @@ trait ApplicationController extends BaseController with GatekeeperAuthWrapper {
     }
   }
 
+  def createPrivOrROPCApplicationPage(): Action[AnyContent] = { requiresRole(Role.APIGatekeeper, requiresSuperUser = true) {
+    implicit request =>
+      implicit hc => {
+        Future.successful(Ok(createPrivOrROPCApplication(createPrivOrROPCAppForm.fill(CreatePrivOrROPCAppForm(None, "", "", "")))))
```
Comment (peteslater-ee) [inline app/controllers/ApplicationController.scala]: Could change case class to have default values and avoid the empty strings here.

File: app/controllers/ApplicationController.scala [hunk 3]
```diff
@@ -387,4 +387,58 @@ trait ApplicationController extends BaseController with GatekeeperAuthWrapper {
...
+              }
+            }
+          }
+
+          def handleFormWithInvalidName: Future[Result] = {
+            val formWithErrors = CreatePrivOrROPCAppForm.invalidAppName(createPrivOrROPCAppForm.fill(form))
+            Future.successful(BadRequest(createPrivOrROPCApplication(formWithErrors)))
+          }
+
+          def hasValidName(apps: Seq[ApplicationResponse]) = {
```
Comment (peteslater-ee) [inline app/controllers/ApplicationController.scala]: This function is a little smelly - could we tidy it up a little bit?
The magic "PRODUCTION" strings feel wrong.

Maybe...

```
def hasValidName(apps: Seq[ApplicationResponse]) = env match {
            case Environment.PRODUCTION => !apps.exists(app => (app.deployedTo == Environment.PRODUCTION.toString) && (app.name == form.applicationName))
            case _ => true
          }
```

File: conf/app.routes [hunk 4]
```diff
@@ -43,4 +43,7 @@ GET         /assets/*file                                           controllers.
 GET         /nav-links                                              controllers.NavigationController.navLinks
 
 GET         /applications/:id/delete                                controllers.ApplicationController.deleteApplicationPage(id: String)
-POST        /applications/:id/delete                                controllers.ApplicationController.deleteApplicationAction(id: String)
\ No newline at end of file
+POST        /applications/:id/delete                                controllers.ApplicationController.deleteApplicationAction(id: String)
+
+GET         /create-priv-or-ROPC-app                                controllers.ApplicationController.createPrivOrROPCApplicationPage
```
Comment (peteslater-ee) [inline conf/app.routes]: These routes could be a little nicer.
Maybe `/application/create` or similar?

File: conf/messages [hunk 5]
```diff
@@ -6,4 +6,10 @@ scopes.required=Scopes are required
 invalid.scope=You have used a scope we don''t recognise
 application.administrator.missing=Choose an administrator
 application.confirmation.missing=Provide the application name
-application.confirmation.error=The application name doesn''t match
\ No newline at end of file
+application.confirmation.error=The application name doesn''t match
```
Comment (peteslater-ee) [inline conf/messages]: Contraction :scream:

File: test/unit/view/applications/CreatePrivOrROPCAppSuccessViewSpec.scala [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        () => html.applications.createPrivOrROPCApplicationSuccess(appId, appName, env, accessType, totpSecret, clientId)(Some(""), applicationMessages, AppConfig)
+
+      page().contentType should include("text/html")
+
+      val document = Jsoup.parse(page().body)
+
+      elementExistsByText(document, "h1", appName) shouldBe true
+      elementExistsByText(document, "h1", "Application added") shouldBe true
+      document.body().toString.contains("This is your only chance to copy and save this application's TOTP secret.") shouldBe true
+      elementExistsByText(document, "tr", "Application ID" + " " + appId) shouldBe true
```
Comment (peteslater-ee) [inline test/unit/view/applications/CreatePrivOrROPCAppSuccessViewSpec.scala]: Use string interpolation here

File: conf/messages (hunk 5 above)
Comment (rose-garratt) [reply conf/messages]: This is a pre-existing message.

**Comment by peteslater-ee** [comment]
The files `createPrivOrROPCApplication.scala.html` and `createPrivOrROPCApplicationSuccess.scala.html` are both named in a camel casey fashion whereas the rest of the view templates are using underscores to separate the words.

Should we be consistent?

**Pull request merged**